numpy/core/src/multiarray/descriptor.c

```diff
@@ -509,8 +509,12 @@ _convert_from_array_descr(PyObject *obj, int align)
                  && (PyUString_Check(title) || PyUnicode_Check(title))
 #endif
                  && (PyDict_GetItem(fields, title) != NULL))) {
-            PyErr_SetString(PyExc_ValueError,
-                    "two fields with the same name");
+#if defined(NPY_PY3K)
+            name = PyUnicode_AsUTF8String(name);
+            Py_DECREF(name);
```

Comment:
This will not work: once you decref `next`, it could be garbage collected before you call `PyString_AsString` on it. What I had in mind on my other comment was something like:

```
#if defined(NPY_PY3K)
            name = PyUnicode_AsUTF8String(name);
#endif
            PyErr_Format(PyExc_ValueError,
                    "field '%s' occurs more than once", PyString_AsString(name));
#if defined(NPY_PY3K)
            Py_DECREF(next);
#endif
            goto fail;
```

It ain't pretty, but neither was the previous version, and at least it puts all the weirdness in the Python 3 code path, and leaves the Python 2 one nice and clean.